python/lsst/rapid/analysis/butlerUtils.py

```diff
@@ -0,0 +1,682 @@
+    return sorted(set([r.day_obs for r in records]))
+
+
+def getMostRecentDataId(butler):
+    """Get the dataId for the most recent observation.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler
+        The butler to query.
+
+    Returns
+    -------
+    dataId : `dict`
+        The dataId of the most recent exposure.
+    """
+    lastDay = getMostRecentDayObs(butler)
+    seqNum = getSeqNumsForDayObs(butler, lastDay)[-1]
+    dataId = {'day_obs': lastDay, 'seq_num': seqNum, 'detector': 0}
+    dataId.update(getExpIdFromDayObsSeqNum(butler, dataId))
+    return dataId
+
+
+def getExpIdFromDayObsSeqNum(butler, dataId):
+    """Get the exposure id for the dataId.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler
+        The butler to query.
+
+    dataId : `dict`
+        The dataId for which to return the exposure id.
+
+    Returns
+    -------
+    dataId : `dict`
+        The dataId of the most recent exposure.
+    """
+    expRecord = getExpRecordFromDataId(butler, dataId)
+    return {'exposure': expRecord.id}
+
+
+def updateDataIdOrDataCord(dataId, **updateKwargs):
+    """Add key, value pairs to a dataId or data coordinate.
+
+    Parameters
+    ----------
+    dataId : `dict`
+        The dataId for which to return the exposure id.
+
+    updateKwargs : `dict`
+        The key value pairs add to the dataId or dataCoord.
+
+    Returns
+    -------
+    dataId : `dict`
+        The updated dataId.
+
+    Notes
+    -----
+    Always returns a dict, so note that if a data coordinate is supplied, a
+    dict is returned, changing the type.
+    """
+    newId = copy.copy(dataId)
+    newId = _assureDict(newId)
+    newId.update(updateKwargs)
+    return newId
+
+
+def fillDataId(butler, dataId):
+    """Given a dataId, fill it with values for all available dimensions.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+
+    dataId : `dict`
+        The dataId to fill.
+
+    Returns
+    -------
+    dataId : `dict`
+        The filled dataId.
+
+    Notes
+    -----
+    This function is *slow*! Running this on 20,000 dataIds takes approximately
+    7 minutes. Virtually all the slowdown is in the
+    butler.registry.expandDataId() call though, so this wrapper is not to blame
+    here, and might speed up in future with butler improvements.
+    """
+    # ensure it's a dict to deal with records etc
+    dataId = _assureDict(dataId)
+
+    # this removes extraneous keys that would trip up the registry call
+    # using _rewrite_data_id is perhaps ever so slightly slower than popping
+    # the bad keys, or making a minimal dataId by hand, but is more
+    # reliable/general, so we choose that over the other approach here
+    dataId, _ = butler._rewrite_data_id(dataId, butler.registry.getDatasetType('raw'))
+
+    # now expand and turn back to a dict
+    dataId = butler.registry.expandDataId(dataId, detector=0).full  # this call is VERY slow
```

Comment:
This is a general comment to middleware, right, and there's nothing for me to actually do here, right? Just checking I'm not missing an action here.

Reply:
I was mainly responding to your comment about this being very slow. If there is ever a place where you are dealing with multiple queryDatasets results then it would be better to call `.expanded()` directly on the result rather than effectively calling a loop with expandDataId.

Additionally, the thing returned by `.expandDataId` will have the dimension record attached to it. The code below seems to be trying to get the dayObs and seqNum from a dataId but you already have it because expandDataId is giving you all possible information (which is why it's slow). If the dataId has been fully expanded then you should have the exposure and day_obs/seq_num (albeit the latter two in a different place). Converting to a dict too soon is making extra work for you.